onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -31,18 +33,17 @@ namespace onnxruntime {
     }                                                           \
   } while (0)
 
-TensorrtExecutionProvider::TensorrtExecutionProvider()
-    : IExecutionProvider{onnxruntime::kTensorrtExecutionProvider} {
-  DeviceAllocatorRegistrationInfo trt_device_info({OrtMemTypeCPU, [](int) {
-                                                     return std::make_unique<TensorrtPinnedAllocator>();
-                                                   },
-                                                   std::numeric_limits<size_t>::max()});
-  InsertAllocator(CreateAllocator(trt_device_info));
-  DeviceAllocatorRegistrationInfo default_device_info({OrtMemTypeDefault, [](int) {
-                                                         return std::make_unique<TensorrtAllocator>();
-                                                       },
-                                                       std::numeric_limits<size_t>::max()});
-  InsertAllocator(CreateAllocator(default_device_info));
+TensorrtExecutionProvider::TensorrtExecutionProvider(const TensorrtExecutionProviderInfo& info)
+    : IExecutionProvider{onnxruntime::kTensorrtExecutionProvider}, device_id_(info.device_id) {
+  CUDA_CALL_THROW(cudaSetDevice(device_id_));
+
+  DeviceAllocatorRegistrationInfo default_allocator_info(
+      {OrtMemTypeDefault, [](int id) { return std::make_unique<TensorrtAllocator>(id); }, std::numeric_limits<size_t>::max()});
+  InsertAllocator(CreateAllocator(default_allocator_info, device_id_));
+
+  DeviceAllocatorRegistrationInfo pinned_allocator_info(
```

Comment:
>pinned_allocator_info [](start = 34, length = 21)

why need the pinned memory for TRT?

Reply:
This allocator is just to let planner get allocator information.

Reply:
I do not understand, pinned memory is not used for trt. For cuda, it's used for asyncbuffer

---
In reply to: [311861315](https://github.com/microsoft/onnxruntime/pull/1561#discussion_r311861315) [](ancestors = 311861315)

Reply:
Yes. Pinned memory is not actually used in TRT. I discussed this with Chen before and it seems we need to register both to let planner get allocator info.